### PR TITLE
Preserve a function's docstring with copy_argspec

### LINKED ISSUE #422: Composite decorator doesn't preserve docstring
Seems like a nice to have at least when creating a lib of strategies for others to use

### PR DESCRIPTION
**Pull request by alexwlchan**
This means that when a strategy is wrapped with `@composite`, we
preserve the docstring of the original function.

Fixes #422.

### CODE REVIEW

File: src/hypothesis/internal/reflection.py [hunk 1]
```diff
@@ -360,14 +360,15 @@ def source_exec_as_module(source):
...
+        '''%(docstring)s'''
         return %(funcname)s(%(invocation)s)
     return %(name)s
 """.strip() + '\n'
 
 
-def copy_argspec(name, argspec):
-    """A decorator which sets the name and argspec of the function passed into
-    it."""
+def copy_argspec(name, docstring, argspec):
```
Comment (DRMacIver) [inline src/hypothesis/internal/reflection.py]: I'm going to nitpick about naming, sorry: The patch looks fine, but I feel like this is significant scope creep for a function named copy\_argspec.

Given that it's an internal function that is used only in a handful of places I think the best thing to do is probably to rename it. Maybe define\_function\_signature or something suitably generic?

**Comment by alexwlchan** [comment]
Changed the name to something more generic.

File: src/hypothesis/internal/reflection.py [hunk 2]
```diff
@@ -360,14 +360,15 @@ def source_exec_as_module(source):
 
 def accept(%(funcname)s):
     def %(name)s(%(argspec)s):
+        '''%(docstring)s'''
```
Comment (DRMacIver) [inline src/hypothesis/internal/reflection.py]: Sorry, I should have caught this before.

I don't think this is a good way to do this and think it'll almost certainly get things mangled. Assigning the \_\_doc\_\_ property works (I think we do that elsewhere?) and should probably be done instead.

Comment (alexwlchan) [inline src/hypothesis/internal/reflection.py]: Fixed, I think.

**Comment by DRMacIver** [comment]
Can you rebase this from master? Otherwise I think it's good to go.

**Comment by alexwlchan** [comment]
> Can you rebase this from master? 

Yep, will do when I finish work.

**Comment by alexwlchan** [comment]
Okay, rebased correctly.

**Comment by alexwlchan** [comment]
Rebased against master; I’ll merge once the tests pass.
